producer.go

```diff
@@ -106,48 +107,73 @@ type AsyncApexLogNSQHandler struct {
 // be published to.
 //
 func NewAsyncApexLogNSQHandler(marshalFunc MarshalFunc, publishFunc PublishFunc, topic string, bufferSize int) *AsyncApexLogNSQHandler {
+	var backoff = 0
 	logChan := make(chan *log.Entry, bufferSize)
 	stopChan := make(chan bool, 1)
+
+	incrementBackoff := func(backoff int) int {
+		if backoff == 0 {
+			return 1
+		}
+		newBackoff := backoff * 2
+		if newBackoff > maximumBackoffMultiple {
+			return maximumBackoffMultiple
+		}
+		return newBackoff
+	}
+
+	handler := &AsyncApexLogNSQHandler{
+		logChan:  logChan,
+		stopChan: stopChan,
+	}
+
 	go func(cLog chan *log.Entry, cStop chan bool) {
 		var e *log.Entry
 		for {
+			if backoff > 0 {
+				time.Sleep(time.Second * time.Duration(backoff))
+			}
 			select {
 			case e = <-cLog:
 				payload, err := marshalFunc(e)
 				if err != nil {
+					handler.mu.Lock()
 					backupLogger.WithError(err).Error("Marshalling log.Entry in AsyncApexLogNSQHander")
-					backupLogger.Handler.HandleLog(e)
+					backoff = incrementBackoff(backoff)
+					handler.mu.Unlock()
+
 					continue
 				}
+
 				err = publishFunc(topic, payload)
 				if err != nil {
-					backupLogger.WithError(err).Error("Publishing in AsyncApexLogNSQHander")
-					backupLogger.Handler.HandleLog(e)
+					handler.mu.Lock()
+					backupLogger.WithField("backoff", time.Second*time.Duration(backoff)).WithError(err).Error("Publishing in AsyncApexLogNSQHander")
+					backoff = incrementBackoff(backoff)
+					handler.mu.Unlock()
+
 					continue
 				}
+				backoff = 0
 			case <-cStop:
 				break
```

Comment:
the reason you are seeing race conditions even after calling `handler.Close()` is due to this break clause only breaking the select statement, not the outer loop. if you changed this to a return, or used a labeled loop/break you wouldn't need locking on the backupLogger calls

https://play.golang.org/p/p7tHx8JjFSY

fixed:

https://play.golang.org/p/FO8AGatNq5Y

Reply:
Done.